# 📤 Import Libraries

In [1]:
# Install a conda package in the current Jupyter kernel
import sys
# !conda install --yes --prefix {sys.prefix} hvplot
!conda install --yes --prefix {sys.prefix} pycaret-nightly


import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
# import hvplot.pandas
%matplotlib inline

Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with initial frozen solve. Retrying with flexible solve.

PackagesNotFoundError: The following packages are not available from current channels:

  - pycaret-nightly

Current channels:

  - https://repo.anaconda.com/pkgs/main/osx-64
  - https://repo.anaconda.com/pkgs/main/noarch
  - https://repo.anaconda.com/pkgs/r/osx-64
  - https://repo.anaconda.com/pkgs/r/noarch

To search for alternate channels that may provide the conda package you're
looking for, navigate to

    https://anaconda.org

and use the search bar at the top of the page.




## 💾 Check out the Data

In [2]:
TRAIN = pd.read_csv('cosc2789-2021/cleaned_train.csv')
VALIDATION = pd.read_csv('cosc2789-2021/cleaned_validation.csv')

TEST = pd.read_csv('cosc2789-2021/cleaned_test.csv')
TRAIN.head()

Severity  Start_Lat  Start_Lng   End_Lat   End_Lng  Distance(mi) Side  \
0         2   39.10148  -84.52341  39.09841 -84.52241         0.219    r   
1         2   41.06213  -81.53784  41.06217 -81.53547         0.123    r   
2         2   39.06324  -84.03243  39.06731 -84.05851         1.427    r   
3         2   40.10931  -82.96849  40.11078 -82.98400         0.826    r   
4         2   40.99613  -85.26613  40.98982 -85.26986         0.477    r   

           City    County State  ... Astronomical_Twilight Start_Time_Month  \
0    cincinnati  hamilton    oh  ...                   day                2   
1         akron    summit    oh  ...                   day                2   
2  williamsburg  clermont    oh  ...                   day                2   
3   westerville  franklin    oh  ...                   day                2   
4       roanoke     allen    in  ...                 night                2   

  Start_Time_Year  Start_Time_Hour  End_Time_Month  End_Time_Year  \
0            2016                6               2           2016   
1            2016                6               2           2016   
2            2016                8               2           2016   
3            2016               15               2           2016   
4            2016               21               2           2016   

   End_Time_Hour Weather_Timestamp_Month  Weather_Timestamp_Year  \
0             12                       2                    2016   
1             12                       2                    2016   
2             14                       2                    2016   
3             21                       2                    2016   
4              3                       2                    2016   

   Weather_Time_Hour  
0                  5  
1                  6  
2                  8  
3                 15  
4                 20  

[5 rows x 46 columns]

In [3]:
TRAIN.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 438700 entries, 0 to 438699
Data columns (total 46 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Severity                 438700 non-null  int64  
 1   Start_Lat                438700 non-null  float64
 2   Start_Lng                438700 non-null  float64
 3   End_Lat                  438700 non-null  float64
 4   End_Lng                  438700 non-null  float64
 5   Distance(mi)             438700 non-null  float64
 6   Side                     438700 non-null  object 
 7   City                     438700 non-null  object 
 8   County                   438700 non-null  object 
 9   State                    438700 non-null  object 
 10  Zipcode                  438700 non-null  object 
 11  Timezone                 438700 non-null  object 
 12  Airport_Code             438700 non-null  object 
 13  Temperature(F)           438700 non-null  float64
 14  Humi

In [4]:
TRAIN.describe()

Severity      Start_Lat      Start_Lng        End_Lat  \
count  438700.000000  438700.000000  438700.000000  438700.000000   
mean        2.243615      36.924992     -98.719154      36.925070   
std         0.612126       5.156410      18.477496       5.156403   
min         1.000000      24.570222    -124.497567      24.570110   
25%         2.000000      33.858560    -118.214156      33.858335   
50%         2.000000      37.354780     -94.525898      37.354760   
75%         2.000000      40.732080     -80.940801      40.732111   
max         4.000000      49.000560     -67.403551      49.000760   

             End_Lng   Distance(mi)  Temperature(F)    Humidity(%)  \
count  438700.000000  438700.000000   438700.000000  438700.000000   
mean      -98.719007       0.586301       59.671724      64.663488   
std        18.477345       1.630518       18.140203      22.951424   
min      -124.497478       0.000000      -89.000000       1.000000   
25%      -118.214103       0.000000       48.000000      49.000000   
50%       -94.525877       0.178000       61.000000      68.000000   
75%       -80.940395       0.593000       73.000000      84.000000   
max       -67.403550     150.138000      168.800000     100.000000   

        Pressure(in)  Visibility(mi)  ...  Precipitation(in)  \
count  438700.000000   438700.000000  ...      438700.000000   
mean       29.568323        9.161630  ...           0.005713   
std         0.999707        2.861522  ...           0.104330   
min         0.300000        0.060000  ...           0.000000   
25%        29.470000       10.000000  ...           0.000000   
50%        29.880000       10.000000  ...           0.000000   
75%        30.040000       10.000000  ...           0.000000   
max        38.940000      120.000000  ...          24.000000   

       Start_Time_Month  Start_Time_Year  Start_Time_Hour  End_Time_Month  \
count     438700.000000    438700.000000    438700.000000   438700.000000   
mean           7.681851      2018.894274        12.546825        7.685297   
std            3.649626         1.365008         6.152619        3.651002   
min            1.000000      2016.000000         0.000000        1.000000   
25%            4.000000      2018.000000         8.000000        4.000000   
50%            9.000000      2020.000000        14.000000        9.000000   
75%           11.000000      2020.000000        17.000000       11.000000   
max           12.000000      2020.000000        23.000000       12.000000   

       End_Time_Year  End_Time_Hour  Weather_Timestamp_Month  \
count  438700.000000  438700.000000            438700.000000   
mean     2018.894438      12.717370                 7.681824   
std         1.364907       7.029786                 3.649635   
min      2016.000000       0.000000                 1.000000   
25%      2018.000000       7.000000                 4.000000   
50%      2020.000000      14.000000                 9.000000   
75%      2020.000000      18.000000                11.000000   
max      2021.000000      23.000000                12.000000   

       Weather_Timestamp_Year  Weather_Time_Hour  
count           438700.000000      438700.000000  
mean              2018.894274          12.268894  
std                  1.365008           6.085492  
min               2016.000000           0.000000  
25%               2018.000000           7.000000  
50%               2020.000000          13.000000  
75%               2020.000000          17.000000  
max               2020.000000          23.000000  

[8 rows x 21 columns]

In [5]:
TRAIN.columns

Index(['Severity', 'Start_Lat', 'Start_Lng', 'End_Lat', 'End_Lng',
       'Distance(mi)', 'Side', 'City', 'County', 'State', 'Zipcode',
       'Timezone', 'Airport_Code', 'Temperature(F)', 'Humidity(%)',
       'Pressure(in)', 'Visibility(mi)', 'Wind_Direction', 'Wind_Speed(mph)',
       'Precipitation(in)', 'Weather_Condition', 'Amenity', 'Bump', 'Crossing',
       'Give_Way', 'Junction', 'No_Exit', 'Railway', 'Roundabout', 'Station',
       'Stop', 'Traffic_Calming', 'Traffic_Signal', 'Sunrise_Sunset',
       'Civil_Twilight', 'Nautical_Twilight', 'Astronomical_Twilight',
       'Start_Time_Month', 'Start_Time_Year', 'Start_Time_Hour',
       'End_Time_Month', 'End_Time_Year', 'End_Time_Hour',
       'Weather_Timestamp_Month', 'Weather_Timestamp_Year',
       'Weather_Time_Hour'],
      dtype='object')

In [6]:
TEST.columns

Index(['ID', 'Start_Lat', 'Start_Lng', 'End_Lat', 'End_Lng', 'Distance(mi)',
       'Side', 'City', 'County', 'State', 'Zipcode', 'Timezone',
       'Airport_Code', 'Temperature(F)', 'Humidity(%)', 'Pressure(in)',
       'Visibility(mi)', 'Wind_Direction', 'Wind_Speed(mph)',
       'Precipitation(in)', 'Weather_Condition', 'Amenity', 'Bump', 'Crossing',
       'Give_Way', 'Junction', 'No_Exit', 'Railway', 'Roundabout', 'Station',
       'Stop', 'Traffic_Calming', 'Traffic_Signal', 'Sunrise_Sunset',
       'Civil_Twilight', 'Nautical_Twilight', 'Astronomical_Twilight',
       'Start_Time_Month', 'Start_Time_Year', 'Start_Time_Hour',
       'End_Time_Month', 'End_Time_Year', 'End_Time_Hour',
       'Weather_Timestamp_Month', 'Weather_Timestamp_Year',
       'Weather_Time_Hour'],
      dtype='object')

In [7]:
X = TRAIN[['Distance(mi)', 'Temperature(F)', 'Humidity(%)', 'Pressure(in)', 'Visibility(mi)', 'Wind_Speed(mph)', 'Precipitation(in)','End_Time_Year']]
y = TRAIN['Severity']

## 🧱 Train Test Split

Now let's split the data into a training set and a testing set. We will train out model on the training set and then use the test set to evaluate the model.

In [8]:
from sklearn.model_selection import train_test_split
#
X_train = X
y_train = y

accident_ID = TEST.ID.to_list()
TEST = TEST[['Distance(mi)', 'Temperature(F)', 'Humidity(%)', 'Pressure(in)', 'Visibility(mi)', 'Wind_Speed(mph)', 'Precipitation(in)','End_Time_Year']]

X_test = VALIDATION[['Distance(mi)', 'Temperature(F)', 'Humidity(%)', 'Pressure(in)', 'Visibility(mi)', 'Wind_Speed(mph)', 'Precipitation(in)','End_Time_Year']]
y_test = VALIDATION[['Severity']]

In [9]:
from sklearn import metrics
from sklearn.model_selection import cross_val_score

def cross_val(model):
    pred = cross_val_score(model, X, y, cv=10)
    return pred.mean()

def print_evaluate(true, predicted):  
    mae = metrics.mean_absolute_error(true, predicted)
    mse = metrics.mean_squared_error(true, predicted)
    rmse = np.sqrt(metrics.mean_squared_error(true, predicted))
    r2_square = metrics.r2_score(true, predicted)
    print('MAE:', mae)
    print('MSE:', mse)
    print('RMSE:', rmse)
    print('R2 Square', r2_square)
    print('__________________________________')
    
def evaluate(true, predicted):
    mae = metrics.mean_absolute_error(true, predicted)
    mse = metrics.mean_squared_error(true, predicted)
    rmse = np.sqrt(metrics.mean_squared_error(true, predicted))
    r2_square = metrics.r2_score(true, predicted)
    return mae, mse, rmse, r2_square

# 📦 Preparing Data For Linear Regression
> Linear regression is been studied at great length, and there is a lot of literature on how your data must be structured to make best use of the model.

> As such, there is a lot of sophistication when talking about these requirements and expectations which can be intimidating. In practice, you can uses these rules more as rules of thumb when using Ordinary Least Squares Regression, the most common implementation of linear regression.

> Try different preparations of your data using these heuristics and see what works best for your problem.
- **Linear Assumption.** Linear regression assumes that the relationship between your input and output is linear. It does not support anything else. This may be obvious, but it is good to remember when you have a lot of attributes. You may need to transform data to make the relationship linear (e.g. log transform for an exponential relationship).
- **Remove Noise.** Linear regression assumes that your input and output variables are not noisy. Consider using data cleaning operations that let you better expose and clarify the signal in your data. This is most important for the output variable and you want to remove outliers in the output variable (y) if possible.
- **Remove Collinearity.** Linear regression will over-fit your data when you have highly correlated input variables. Consider calculating pairwise correlations for your input data and removing the most correlated.
- **Gaussian Distributions.** Linear regression will make more reliable predictions if your input and output variables have a Gaussian distribution. You may get some benefit using transforms (e.g. log or BoxCox) on you variables to make their distribution more Gaussian looking.
- **Rescale Inputs:** Linear regression will often make more reliable predictions if you rescale input variables using standardization or normalization.

In [10]:
from sklearn.preprocessing import StandardScaler
from sklearn.pipeline import Pipeline

pipeline = Pipeline([
    ('std_scalar', StandardScaler())
])

X_train = pipeline.fit_transform(X_train)
X_test = pipeline.transform(X_test)

# ✔️ Linear Regression

In [11]:
from sklearn.linear_model import LinearRegression

lin_reg = LinearRegression(normalize=True)
lin_reg.fit(X_train,y_train)

LinearRegression(normalize=True)

## ✔️ Model Evaluation

Let's evaluate the model by checking out it's coefficients and how we can interpret them.

In [12]:
# print the intercept
print(lin_reg.intercept_)

2.2436152268064675


In [13]:
coeff_df = pd.DataFrame(lin_reg.coef_, X.columns, columns=['Coefficient'])
coeff_df

Coefficient
Distance(mi)          0.087860
Temperature(F)       -0.014064
Humidity(%)           0.034981
Pressure(in)         -0.004078
Visibility(mi)        0.014190
Wind_Speed(mph)       0.023197
Precipitation(in)    -0.000998
End_Time_Year        -0.138812

> Interpreting the coefficients:
- Holding all other features fixed, a 1 unit increase in **Distance(mi)** is associated with an **increase of \0.087860**.
- Holding all other features fixed, a 1 unit increase in **Temperature(F)** is associated with an **decrease of \-0.014064**.

## ✔️ Predictions from our Model

Let's grab predictions off our test set and see how well it did!

In [14]:
pred = lin_reg.predict(X_test)

In [ ]:
# import hvplot.pandas

# pd.DataFrame({'True Values': y_test, 'Predicted Values': pred}).hvplot.scatter(x='True Values', y='Predicted Values')

**Residual Histogram**

In [16]:
# pd.DataFrame({'Error Values': (y_test - pred)}).hvplot.kde()

## ✔️ Regression Evaluation Metrics


Here are three common evaluation metrics for regression problems:

> - **Mean Absolute Error** (MAE) is the mean of the absolute value of the errors:
$$\frac 1n\sum_{i=1}^n|y_i-\hat{y}_i|$$

> - **Mean Squared Error** (MSE) is the mean of the squared errors:
$$\frac 1n\sum_{i=1}^n(y_i-\hat{y}_i)^2$$

> - **Root Mean Squared Error** (RMSE) is the square root of the mean of the squared errors:
$$\sqrt{\frac 1n\sum_{i=1}^n(y_i-\hat{y}_i)^2}$$

> 📌 Comparing these metrics:
- **MAE** is the easiest to understand, because it's the average error.
- **MSE** is more popular than MAE, because MSE "punishes" larger errors, which tends to be useful in the real world.
- **RMSE** is even more popular than MSE, because RMSE is interpretable in the "y" units.

> All of these are **loss functions**, because we want to minimize them.

In [17]:
test_pred = lin_reg.predict(X_test)
train_pred = lin_reg.predict(X_train)

print('Test set evaluation:\n_____________________________________')
print_evaluate(y_test, test_pred)
print('Train set evaluation:\n_____________________________________')
print_evaluate(y_train, train_pred)

Test set evaluation:
_____________________________________
MAE: 0.396735991419281
MSE: 0.34239345438028196
RMSE: 0.5851439603894771
R2 Square 0.07947902370064608
__________________________________
Train set evaluation:
_____________________________________
MAE: 0.3997139491539972
MSE: 0.34458531889164357
RMSE: 0.5870139000838426
R2 Square 0.08036438496918397
__________________________________


In [18]:
results_df = pd.DataFrame(data=[["Linear Regression", *evaluate(y_test, test_pred) , cross_val(LinearRegression())]], 
                          columns=['Model', 'MAE', 'MSE', 'RMSE', 'R2 Square', "Cross Validation"])
results_df

Model       MAE       MSE      RMSE  R2 Square  \
0  Linear Regression  0.396736  0.342393  0.585144   0.079479   

   Cross Validation  
0         -0.135238

In [19]:
y_predict_model_lin_reg = lin_reg.predict(TEST)

output = pd.DataFrame({"ID": accident_ID, "Severity": y_predict_model_lin_reg})

output.to_csv('submission_arti.csv', index=False)
print("Submission was successfully saved!")

Submission was successfully saved!


# ✔️ Ridge Regression

> Source: [scikit-learn](http://scikit-learn.org/stable/modules/linear_model.html#ridge-regression)

> Ridge regression addresses some of the problems of **Ordinary Least Squares** by imposing a penalty on the size of coefficients. The ridge coefficients minimize a penalized residual sum of squares,

$$\min_{w}\big|\big|Xw-y\big|\big|^2_2+\alpha\big|\big|w\big|\big|^2_2$$

> $\alpha>=0$ is a complexity parameter that controls the amount of shrinkage: the larger the value of $\alpha$, the greater the amount of shrinkage and thus the coefficients become more robust to collinearity.

> Ridge regression is an L2 penalized model. Add the squared sum of the weights to the least-squares cost function.
***

In [20]:
from sklearn.linear_model import Ridge

model = Ridge(alpha=100, solver='cholesky', tol=0.0001, random_state=42)
model.fit(X_train, y_train)
pred = model.predict(X_test)

test_pred = model.predict(X_test)
train_pred = model.predict(X_train)

print('Test set evaluation:\n_____________________________________')
print_evaluate(y_test, test_pred)
print('====================================')
print('Train set evaluation:\n_____________________________________')
print_evaluate(y_train, train_pred)

Test set evaluation:
_____________________________________
MAE: 0.3967432596435439
MSE: 0.34239315777208296
RMSE: 0.5851437069405113
R2 Square 0.07947982112847496
__________________________________
Train set evaluation:
_____________________________________
MAE: 0.3997210355530711
MSE: 0.34458532046786255
RMSE: 0.5870139014264164
R2 Square 0.08036438076254171
__________________________________


In [21]:
results_df_2 = pd.DataFrame(data=[["Ridge Regression", *evaluate(y_test, test_pred) , cross_val(Ridge())]], 
                            columns=['Model', 'MAE', 'MSE', 'RMSE', 'R2 Square', "Cross Validation"])
results_df = results_df.append(results_df_2, ignore_index=True)
results_df

Model       MAE       MSE      RMSE  R2 Square  \
0  Linear Regression  0.396736  0.342393  0.585144   0.079479   
1   Ridge Regression  0.396743  0.342393  0.585144   0.079480   

   Cross Validation  
0         -0.135238  
1         -0.135238

# ✔️ LASSO Regression

> A linear model that estimates sparse coefficients.

> Mathematically, it consists of a linear model trained with $\ell_1$ prior as regularizer. The objective function to minimize is:

$$\min_{w}\frac{1}{2n_{samples}} \big|\big|Xw - y\big|\big|_2^2 + \alpha \big|\big|w\big|\big|_1$$

> The lasso estimate thus solves the minimization of the least-squares penalty with $\alpha \big|\big|w\big|\big|_1$ added, where $\alpha$ is a constant and $\big|\big|w\big|\big|_1$ is the $\ell_1-norm$ of the parameter vector.
***

In [22]:
from sklearn.linear_model import Lasso

model = Lasso(alpha=0.1, 
              precompute=True, 
#               warm_start=True, 
              positive=True, 
              selection='random',
              random_state=42)
model.fit(X_train, y_train)

test_pred = model.predict(X_test)
train_pred = model.predict(X_train)

print('Test set evaluation:\n_____________________________________')
print_evaluate(y_test, test_pred)
print('====================================')
print('Train set evaluation:\n_____________________________________')
print_evaluate(y_train, train_pred)

Test set evaluation:
_____________________________________
MAE: 0.43189534125318396
MSE: 0.37195711259002145
RMSE: 0.6098828679263104
R2 Square -2.5410607653597594e-06
__________________________________
Train set evaluation:
_____________________________________
MAE: 0.4338350159559747
MSE: 0.3746976664013788
RMSE: 0.6121255315712446
R2 Square 0.0
__________________________________


In [23]:
results_df_2 = pd.DataFrame(data=[["Lasso Regression", *evaluate(y_test, test_pred) , cross_val(Lasso())]], 
                            columns=['Model', 'MAE', 'MSE', 'RMSE', 'R2 Square', "Cross Validation"])
results_df = results_df.append(results_df_2, ignore_index=True)
results_df

Model       MAE       MSE      RMSE  R2 Square  \
0  Linear Regression  0.396736  0.342393  0.585144   0.079479   
1   Ridge Regression  0.396743  0.342393  0.585144   0.079480   
2   Lasso Regression  0.431895  0.371957  0.609883  -0.000003   

   Cross Validation  
0         -0.135238  
1         -0.135238  
2         -0.283739

# ✔️ Elastic Net

> A linear regression model trained with L1 and L2 prior as regularizer. 

> This combination allows for learning a sparse model where few of the weights are non-zero like Lasso, while still maintaining the regularization properties of Ridge. 

> Elastic-net is useful when there are multiple features which are correlated with one another. Lasso is likely to pick one of these at random, while elastic-net is likely to pick both.

> A practical advantage of trading-off between Lasso and Ridge is it allows Elastic-Net to inherit some of Ridge’s stability under rotation.

> The objective function to minimize is in this case

$$\min_{w}{\frac{1}{2n_{samples}} \big|\big|X w - y\big|\big|_2 ^ 2 + \alpha \rho \big|\big|w\big|\big|_1 +
\frac{\alpha(1-\rho)}{2} \big|\big|w\big|\big|_2 ^ 2}$$
***

In [24]:
from sklearn.linear_model import ElasticNet

model = ElasticNet(alpha=0.1, l1_ratio=0.9, selection='random', random_state=42)
model.fit(X_train, y_train)

test_pred = model.predict(X_test)
train_pred = model.predict(X_train)

print('Test set evaluation:\n_____________________________________')
print_evaluate(y_test, test_pred)
print('====================================')
print('Train set evaluation:\n_____________________________________')
print_evaluate(y_train, train_pred)

Test set evaluation:
_____________________________________
MAE: 0.4219970856318469
MSE: 0.35936427959007833
RMSE: 0.5994699989074335
R2 Square 0.03385314975103282
__________________________________
Train set evaluation:
_____________________________________
MAE: 0.42438154469890454
MSE: 0.36205831914287834
RMSE: 0.6017128211554731
R2 Square 0.03373212163259409
__________________________________


In [25]:
results_df_2 = pd.DataFrame(data=[["Elastic Net Regression", *evaluate(y_test, test_pred) , cross_val(ElasticNet())]], 
                            columns=['Model', 'MAE', 'MSE', 'RMSE', 'R2 Square', "Cross Validation"])
results_df = results_df.append(results_df_2, ignore_index=True)
results_df

Model       MAE       MSE      RMSE  R2 Square  \
0       Linear Regression  0.396736  0.342393  0.585144   0.079479   
1        Ridge Regression  0.396743  0.342393  0.585144   0.079480   
2        Lasso Regression  0.431895  0.371957  0.609883  -0.000003   
3  Elastic Net Regression  0.421997  0.359364  0.599470   0.033853   

   Cross Validation  
0         -0.135238  
1         -0.135238  
2         -0.283739  
3         -0.290481

# ✔️ Polynomial Regression
> Source: [scikit-learn](http://scikit-learn.org/stable/modules/linear_model.html#polynomial-regression-extending-linear-models-with-basis-functions)

***

> One common pattern within machine learning is to use linear models trained on nonlinear functions of the data. This approach maintains the generally fast performance of linear methods, while allowing them to fit a much wider range of data.

> For example, a simple linear regression can be extended by constructing polynomial features from the coefficients. In the standard linear regression case, you might have a model that looks like this for two-dimensional data:

$$\hat{y}(w, x) = w_0 + w_1 x_1 + w_2 x_2$$

> If we want to fit a paraboloid to the data instead of a plane, we can combine the features in second-order polynomials, so that the model looks like this:

$$\hat{y}(w, x) = w_0 + w_1 x_1 + w_2 x_2 + w_3 x_1 x_2 + w_4 x_1^2 + w_5 x_2^2$$

> The (sometimes surprising) observation is that this is still a linear model: to see this, imagine creating a new variable

$$z = [x_1, x_2, x_1 x_2, x_1^2, x_2^2]$$

> With this re-labeling of the data, our problem can be written

$$\hat{y}(w, x) = w_0 + w_1 z_1 + w_2 z_2 + w_3 z_3 + w_4 z_4 + w_5 z_5$$

> We see that the resulting polynomial regression is in the same class of linear models we’d considered above (i.e. the model is linear in w) and can be solved by the same techniques. By considering linear fits within a higher-dimensional space built with these basis functions, the model has the flexibility to fit a much broader range of data.
***

In [26]:
from sklearn.preprocessing import PolynomialFeatures

poly_reg = PolynomialFeatures(degree=2)

X_train_2_d = poly_reg.fit_transform(X_train)
X_test_2_d = poly_reg.transform(X_test)

lin_reg = LinearRegression(normalize=True)
lin_reg.fit(X_train_2_d,y_train)

test_pred = lin_reg.predict(X_test_2_d)
train_pred = lin_reg.predict(X_train_2_d)

print('Test set evaluation:\n_____________________________________')
print_evaluate(y_test, test_pred)
print('====================================')
print('Train set evaluation:\n_____________________________________')
print_evaluate(y_train, train_pred)

Test set evaluation:
_____________________________________
MAE: 0.3832169133814977
MSE: 0.32963692619168977
RMSE: 0.5741401624966588
R2 Square 0.11377480719804622
__________________________________
Train set evaluation:
_____________________________________
MAE: 0.3867504793758513
MSE: 0.33156552737604666
RMSE: 0.5758172690846001
R2 Square 0.11511184320835532
__________________________________


In [27]:
results_df_2 = pd.DataFrame(data=[["Polynomail Regression", *evaluate(y_test, test_pred), 0]], 
                            columns=['Model', 'MAE', 'MSE', 'RMSE', 'R2 Square', 'Cross Validation'])
results_df = results_df.append(results_df_2, ignore_index=True)
results_df

Model       MAE       MSE      RMSE  R2 Square  \
0       Linear Regression  0.396736  0.342393  0.585144   0.079479   
1        Ridge Regression  0.396743  0.342393  0.585144   0.079480   
2        Lasso Regression  0.431895  0.371957  0.609883  -0.000003   
3  Elastic Net Regression  0.421997  0.359364  0.599470   0.033853   
4   Polynomail Regression  0.383217  0.329637  0.574140   0.113775   

   Cross Validation  
0         -0.135238  
1         -0.135238  
2         -0.283739  
3         -0.290481  
4          0.000000

# ✔️ Stochastic Gradient Descent

> Gradient Descent is a very generic optimization algorithm capable of finding optimal solutions to a wide range of problems. The general idea of Gradient Sescent is to tweak parameters iteratively in order to minimize a cost function. Gradient Descent measures the local gradient of the error function with regards to the parameters vector, and it goes in the direction of descending gradient. Once the gradient is zero, you have reached a minimum.

In [28]:
from sklearn.linear_model import SGDRegressor

sgd_reg = SGDRegressor(n_iter_no_change=250, penalty=None, eta0=0.0001, max_iter=100000)
sgd_reg.fit(X_train, y_train)

test_pred = sgd_reg.predict(X_test)
train_pred = sgd_reg.predict(X_train)

print('Test set evaluation:\n_____________________________________')
print_evaluate(y_test, test_pred)
print('====================================')
print('Train set evaluation:\n_____________________________________')
print_evaluate(y_train, train_pred)

Test set evaluation:
_____________________________________
MAE: 0.39674640744650885
MSE: 0.3423986446878448
RMSE: 0.5851483954415707
R2 Square 0.07946506961675825
__________________________________
Train set evaluation:
_____________________________________
MAE: 0.39972341017099466
MSE: 0.3445854498244227
RMSE: 0.5870140116082603
R2 Square 0.08036403553338323
__________________________________


In [29]:
results_df_2 = pd.DataFrame(data=[["Stochastic Gradient Descent", *evaluate(y_test, test_pred), 0]], 
                            columns=['Model', 'MAE', 'MSE', 'RMSE', 'R2 Square', 'Cross Validation'])
results_df = results_df.append(results_df_2, ignore_index=True)
results_df

Model       MAE       MSE      RMSE  R2 Square  \
0            Linear Regression  0.396736  0.342393  0.585144   0.079479   
1             Ridge Regression  0.396743  0.342393  0.585144   0.079480   
2             Lasso Regression  0.431895  0.371957  0.609883  -0.000003   
3       Elastic Net Regression  0.421997  0.359364  0.599470   0.033853   
4        Polynomail Regression  0.383217  0.329637  0.574140   0.113775   
5  Stochastic Gradient Descent  0.396746  0.342399  0.585148   0.079465   

   Cross Validation  
0         -0.135238  
1         -0.135238  
2         -0.283739  
3         -0.290481  
4          0.000000  
5          0.000000

In [30]:
y_predict_poly_sgd_reg = sgd_reg.predict(TEST)

output = pd.DataFrame({"ID": accident_ID, "Severity": y_predict_poly_sgd_reg})

output.to_csv('submission_sgd_reg.csv', index=False)
print("Submission was successfully saved!")

Submission was successfully saved!


# ✔️ Artficial Neural Network

In [31]:
# Install a pip package in the current Jupyter kernel
import sys
!{sys.executable} -m pip install tensorflow

from tensorflow.keras.models import Sequential
from tensorflow.keras.layers import Input, Dense, Activation, Dropout
from tensorflow.keras.optimizers import Adam

X_train = np.array(X_train)
X_test = np.array(X_test)
y_train = np.array(y_train)
y_test = np.array(y_test)

model = Sequential()

model.add(Dense(X_train.shape[1], activation='relu'))
model.add(Dense(32, activation='relu'))
# model.add(Dropout(0.2))

model.add(Dense(64, activation='relu'))
# model.add(Dropout(0.2))

model.add(Dense(128, activation='relu'))
# model.add(Dropout(0.2))

model.add(Dense(512, activation='relu'))
model.add(Dropout(0.1))
model.add(Dense(1))

model.compile(optimizer=Adam(0.00001), loss='mse')

r = model.fit(X_train, y_train,
              validation_data=(X_test,y_test),
              batch_size=1,
              epochs=10)

Defaulting to user installation because normal site-packages is not writeable
Epoch 1/10
438700/438700 [==============================] - 47476s 108ms/step - loss: 0.3799 - val_loss: 0.3336
Epoch 2/10
438700/438700 [==============================] - 549s 1ms/step - loss: 0.3362 - val_loss: 0.3280
Epoch 3/10
438700/438700 [==============================] - 511s 1ms/step - loss: 0.3312 - val_loss: 0.3242
Epoch 4/10
438700/438700 [==============================] - 567s 1ms/step - loss: 0.3283 - val_loss: 0.3231
Epoch 5/10
438700/438700 [==============================] - 479s 1ms/step - loss: 0.3265 - val_loss: 0.3208
Epoch 6/10
438700/438700 [==============================] - 566s 1ms/step - loss: 0.3253 - val_loss: 0.3203
Epoch 7/10
438700/438700 [==============================] - 551s 1ms/step - loss: 0.3241 - val_loss: 0.3191
Epoch 8/10
438700/438700 [==============================] - 504s 1ms/step - loss: 0.3232 - val_loss: 0.3186
Epoch 9/10
438700/438700 [============================

In [32]:
# pd.DataFrame({'True Values': y_test, 'Predicted Values': pred}).hvplot.scatter(x='True Values', y='Predicted Values')

In [33]:
# pd.DataFrame(r.history)

In [34]:
# pd.DataFrame(r.history).hvplot.line(y=['loss', 'val_loss'])

In [35]:
test_pred = model.predict(X_test)
train_pred = model.predict(X_train)

print('Test set evaluation:\n_____________________________________')
print_evaluate(y_test, test_pred)

print('Train set evaluation:\n_____________________________________')
print_evaluate(y_train, train_pred)

Test set evaluation:
_____________________________________
MAE: 0.3707921622990742
MSE: 0.31812345525044217
RMSE: 0.5640243392358544
R2 Square 0.14472864517551043
__________________________________
Train set evaluation:
_____________________________________
MAE: 0.37294105291329216
MSE: 0.3185173214888353
RMSE: 0.5643733883599007
R2 Square 0.14993513424330407
__________________________________


In [36]:
results_df_2 = pd.DataFrame(data=[["Artificial Neural Network", *evaluate(y_test, test_pred), 0]],
                            columns=['Model', 'MAE', 'MSE', 'RMSE', 'R2 Square', 'Cross Validation'])
results_df = results_df.append(results_df_2, ignore_index=True)
results_df

Model       MAE       MSE      RMSE  R2 Square  \
0            Linear Regression  0.396736  0.342393  0.585144   0.079479   
1             Ridge Regression  0.396743  0.342393  0.585144   0.079480   
2             Lasso Regression  0.431895  0.371957  0.609883  -0.000003   
3       Elastic Net Regression  0.421997  0.359364  0.599470   0.033853   
4        Polynomail Regression  0.383217  0.329637  0.574140   0.113775   
5  Stochastic Gradient Descent  0.396746  0.342399  0.585148   0.079465   
6    Artificial Neural Network  0.370792  0.318123  0.564024   0.144729   

   Cross Validation  
0         -0.135238  
1         -0.135238  
2         -0.283739  
3         -0.290481  
4          0.000000  
5          0.000000  
6          0.000000

In [ ]:
train_pred

In [37]:
y_predict_model_arti = model.predict(TEST)

output = pd.Series({"ID": accident_ID, "Severity": y_predict_model_arti})

output.to_csv('submission_arti.csv', index=False)
print("Submission was successfully saved!")

Submission was successfully saved!


# ✔️ Random Forest Regressor

In [38]:
from sklearn.ensemble import RandomForestRegressor

rf_reg = RandomForestRegressor(n_estimators=1000)
rf_reg.fit(X_train, y_train)

test_pred = rf_reg.predict(X_test)
train_pred = rf_reg.predict(X_train)

print('Test set evaluation:\n_____________________________________')
print_evaluate(y_test, test_pred)

print('Train set evaluation:\n_____________________________________')
print_evaluate(y_train, train_pred)

Test set evaluation:
_____________________________________
MAE: 0.3449715253030803
MSE: 0.2868093442400351
RMSE: 0.5355458376647466
R2 Square 0.22891628273248066
__________________________________
Train set evaluation:
_____________________________________
MAE: 0.13548375542874475
MSE: 0.04577091879788108
RMSE: 0.21394139103474363
R2 Square 0.8778457329679472
__________________________________


In [39]:
results_df_2 = pd.DataFrame(data=[["Random Forest Regressor", *evaluate(y_test, test_pred), 0]], 
                            columns=['Model', 'MAE', 'MSE', 'RMSE', 'R2 Square', 'Cross Validation'])
results_df = results_df.append(results_df_2, ignore_index=True)
results_df

Model       MAE       MSE      RMSE  R2 Square  \
0            Linear Regression  0.396736  0.342393  0.585144   0.079479   
1             Ridge Regression  0.396743  0.342393  0.585144   0.079480   
2             Lasso Regression  0.431895  0.371957  0.609883  -0.000003   
3       Elastic Net Regression  0.421997  0.359364  0.599470   0.033853   
4        Polynomail Regression  0.383217  0.329637  0.574140   0.113775   
5  Stochastic Gradient Descent  0.396746  0.342399  0.585148   0.079465   
6    Artificial Neural Network  0.370792  0.318123  0.564024   0.144729   
7      Random Forest Regressor  0.344972  0.286809  0.535546   0.228916   

   Cross Validation  
0         -0.135238  
1         -0.135238  
2         -0.283739  
3         -0.290481  
4          0.000000  
5          0.000000  
6          0.000000  
7          0.000000

In [ ]:
y_predict_rf_reg = rf_reg.predict(TEST)

output = pd.DataFrame({"ID": accident_ID, "Severity": y_predict_rf_reg})

output.to_csv('submission_rf_reg.csv', index=False)
print("Submission was successfully saved!")

# ✔️ Support Vector Machine

In [ ]:
from sklearn.svm import SVR

svm_reg = SVR(kernel='rbf', C=1000000, epsilon=0.001)
svm_reg.fit(X_train, y_train)

test_pred = svm_reg.predict(X_test)
train_pred = svm_reg.predict(X_train)

print('Test set evaluation:\n_____________________________________')
print_evaluate(y_test, test_pred)

print('Train set evaluation:\n_____________________________________')
print_evaluate(y_train, train_pred)

In [ ]:
results_df_2 = pd.DataFrame(data=[["SVM Regressor", *evaluate(y_test, test_pred), 0]], 
                            columns=['Model', 'MAE', 'MSE', 'RMSE', 'R2 Square', 'Cross Validation'])
results_df = results_df.append(results_df_2, ignore_index=True)
results_df

In [ ]:
y_predict_svm_reg = svm_reg.predict(TEST)

output = pd.DataFrame({"ID": accident_ID, "Severity": y_predict_svm_reg})

output.to_csv('submission_arti.csv', index=False)
print("Submission was successfully saved!")

# 📊 Models Comparison

In [ ]:
results_df.set_index('Model', inplace=True)
results_df['R2 Square'].plot(kind='barh', figsize=(12, 8))

## Observe TEST predictions

In [ ]:
All = pd.DataFrame({"ID": accident_ID, "Model": y_predict_model_reg, "RandomForest": y_predict_rf_reg, 'SVM':y_predict_svm_reg})
All['Voting']=All[["Model","Ridge",'RandomForest','SVM']].mean(axis=1)
All.head(10)

In [ ]:
disp = 150
fig, ax = plt.subplots(figsize=(30,10))  
for col in All.columns[1:].tolist():
    plt.scatter(x = All[:disp].ID, y= All[:disp][col], alpha=0.8)
    plt.legend(All.columns[1:].tolist())

plt.show()

# 📝 Summary
In this notebook you discovered the linear regression algorithm for machine learning.

You covered a lot of ground including:
> - The common linear regression models (Ridge, Lasso, ElasticNet, ...).
> - The representation used by the model.
> - Learning algorithms used to estimate the coefficients in the model.
> - Rules of thumb to consider when preparing data for use with linear regression.
> - How to evaluate a linear regression model.


# 🔗 References:
- [Scikit-learn library](https://scikit-learn.org/stable/supervised_learning.html#supervised-learning)
- [Linear Regression for Machine Learning by Jason Brownlee PhD](https://machinelearningmastery.com/linear-regression-for-machine-learning/)